api: add keyspace_trigger_offstrategy_compaction

bhalevy opened the pull request:
Add RESTful api for triggering offstrategy compaction
on table(s) in a single keyspace.

Usage: curl -X POST "http://{ip}:10000/storage_service/keyspace_trigger_offstrategy_compaction/{keyspace}?table={table}

With no options, offstrategy compaction is triggered (if pending)
on all tables in the keyspace.

This is useful for triggering offstrategy compaction
post repair, after repairing all token ranges.

Otherwise, offstrategy compaction will only be
auto-triggered after a 5 minutes idle timeout.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

raphaelsc commented on api/storage_service.cc:
```diff
@@ -541,6 +541,25 @@ void set_storage_service(http_context& ctx, routes& r) {
+        return ctx.db.invoke_on_all([keyspace = std::move(keyspace), table_names = std::move(table_names)] (database& db) {
+            for (auto tname : table_names) {
+                auto& t = db.find_column_family(keyspace, tname);
+                if (t.is_pending_offstrategy_compaction()) {
```
It seems to me is_pending_offstrategy_compaction() isn't buying us anything because repair can still be in progress, so offstrategy can be triggered prematurely. 

not sure I like this API as it can be easily misused. efficiency problems will happen if user runs it prematurely, before repair completed. @asias could we add a repair cmd instead which would be sent on completion by user. so offstrategy is triggered, if either the cmd was received, or the timer is triggered, whichever happens first.

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -622,7 +622,7 @@ void compaction_manager::submit(column_family* cf) {
 }
 
-void compaction_manager::submit_offstrategy(column_family* cf) {
+future<> compaction_manager::submit_offstrategy(column_family* cf) {
```
i think we should keep this variant which doesn't return a future.

```
private:
lw_shared_ptr<task> do_submit_offstrategy(column_family*);
public:
future<> perform_offstrategy(column_family*) { 
  auto task = do_submit_offstrategy(cf);
  return task->compaction_done.get_future().finally([task] {});
}
void submit_offstrategy(column_family*) {
  do_submit_offstrategy(cf);
}
```

bhalevy replied:
Fixed in v3

raphaelsc commented on api/storage_service.cc:
```diff
@@ -501,13 +501,19 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
+            return parallel_for_each(column_families_vec, [offstrategy] (column_family* cf) {
+                if (!offstrategy) {
                     return cf->compact_all_sstables();
+                } else if (cf->is_pending_offstrategy_compaction()) {
```
don't think we need cf->is_pending_offstrategy_compaction(). the same way we don't check first if we need major. we just submit a request to compaction manager, which will then execute it while guaranteeing serialization. if there's nothing to be offstrategy compacted, table will handle it just fine.

bhalevy replied:
Fixed in v3

bhalevy commented:
In v3:
- keep `void compaction_manager::submit_offstrategy()`
- expose `future<> compaction_manager::do_submit_offstrategy()`
  - self-contained, checked if offstrategy compaction needs triggering, otherwise returns a ready future.
  - use that from api call via `future<> table::submit_offstrategy_compaction()`

raphaelsc reviewed (approved):
LGTM

bhalevy commented:
@amnonh if you have no objection, @scylladb/scylla-maint let's merge this change, since the following chain of PRs depend on it: scylladb/scylla-jmx#172 -> scylladb/scylla-tools-java#269 -> scylladb/scylla-dtest#2237.

amnonh commented:
LGTM

nyh commented on database.hh:
```diff
@@ -883,6 +883,7 @@ public:
     void try_trigger_compaction() noexcept;
     future<> run_compaction(sstables::compaction_descriptor descriptor);
     void trigger_offstrategy_compaction();
+    future<> submit_offstrategy_compaction();
```
nitpick: It's not great that we have a string of similarly-named functions {trigger,submit,run}_offstrategy_compaction()... Nobody will ever remember what's the difference between all those functions.

bhalevy replied:
Fixed in v4.

nyh commented on table.cc:
```diff
@@ -962,6 +962,17 @@ void table::trigger_offstrategy_compaction() {
+    // If the user calls trigger_offstrategy_compaction() to trigger
+    // off-strategy explicitly, cancel the timeout based automatic trigger.
+    if (!_off_strategy_trigger.armed()) {
+        return make_ready_future<>();
```
Is this logic correct? submit_offstrategy_compaction() should be do absolutely nothing if an offstrategy compaction wasn't already scheduled?
If this is logic is correct, I think at least the name of the function should be different. Perhaps undelay_offstrategy_compaction() or something like that.

raphaelsc replied:
agreed. the function isn't really about submitting offstrategy request, but rather anticipating a offstrategy which would be latter submitted when timer expires. requires a more descriptive name. and important to mention that it's only covering offstrategy for repair. for others there's not a timer.

bhalevy replied:
That's why I liked trigger_offstrategy_compaction (if it's already scheduled, trigger it now)

raphaelsc replied:
agreed. we need a name that describes it better. this function is confined to offstrategy for regular repair, and it will only submit request if there's an ongoing repair.

i tend to think this function should only perform offstrategy compaction, not check for any condition or do anything else. so it could be named run_offstrategy_compaction() instead. if it has such a generic name, it shouldn't be restricted for specific conditions or for a specific maintenance operation.

bhalevy replied:
and the logic is correct, we don't need to start a new offstrategy compaction if the trigger isn't already armed.
If it was never armed, we have nothing to compact offstrategy, If it was armed in the past and not it isn't - it means that offstrategy compaction already completed and it is not needed at the moment.

raphaelsc replied:
trigger_offstrategy_compaction() is confusing if it's going to wait for execution. run_offstrategy_compaction() makes more sense. also if we check for timer, the function is confined for offstrategy on regular repair, then it shouldn't go with such a generic name. it's confusing. 

I'd suggest submit_offstrategy_compaction() to be renamed to run_offstrategy_compaction(). Then remove any condition in the function. It will only call _compaction_manager.perform_offstrategy_compaction(). If we want to make it specific to offstrategy on regular repair, then we better come up with a more descriptive name, and the API should also be documented such that its usage is confined to offstrategy on regular repair.

bhalevy replied:
We can do that.
I'll make sure it will serialize well if the timer expires in the middle of an on-going offstrategy-compaction.
(Another use case that comes into my mind is running offstrategy compaction during repair to limit the backlog of offstrategy compaction.  I'm not sure that accumulating e.g. 1024 sstables is worth it in terms of efficiency, vs. triggering offstrategy compaction periodically when e.g. 32 sstables were accumulated (per shard))

bhalevy replied:
Fixed in v4.

nyh reviewed (commented):
Looks mostly good, but I left a couple of comments for your consideration.

raphaelsc commented on table.cc:
```diff
@@ -962,6 +962,17 @@ void table::trigger_offstrategy_compaction() {
+    }
+
+    _off_strategy_trigger.cancel();
+    return _compaction_manager.do_submit_offstrategy(this);
```
I missed this: do_submit_offstrategy(). functions with this name is usually confirmed to private namespace of a class. given that this is about running / performing offstrategy, I'd like it to have a more descriptive name. Following the name convention of compaction_manager, I  think perform_offstrategy() would be a good one.

bhalevy replied:
Fixed in v4.

bhalevy commented:
In v4:
- `compaction_manager::do_submit_offstrategy` -> `perform_offstrategy`
-  `table::submit_offstrategy_compaction` -> `perform_offstrategy_compaction`
  - do not checkout if _off_strategy_trigger is armed, always call `table::perform_offstrategy_compaction`.
- table: offstrategy compaction: add minimal documentation to api methods

nyh commented on compaction/compaction_manager.cc:
```diff
@@ -623,7 +623,7 @@ void compaction_manager::submit(column_family* cf) {
 }
 
-void compaction_manager::submit_offstrategy(column_family* cf) {
+future<> compaction_manager::perform_offstrategy(column_family* cf) {
```
The commit message still refers to the old names.

nyh commented on api/api-doc/storage_service.json:
```diff
@@ -728,6 +728,14 @@
+                  },
+                  {
+                     "name":"offstrategy",
+                     "description":"When set to true, trigger offstrategy compaction (and return immediately)",
```
If we don't provide a better explanation of what this "offstrategy compaction" means, we need to create documentation about it - Google didn't find any. As a user I would have no idea what exactly will happen when I "trigger offstrategy compaction" or why or when I would want to do it.

bhalevy replied:
I opened https://github.com/scylladb/scylla/issues/9162 as a follow-up.

nyh commented on database.hh:
```diff
@@ -882,9 +882,15 @@ public:
+    // Perform offstrategy compaction
     future<> perform_offstrategy_compaction();
+    // Run offstrategy compaction.
+    // Typically called only internally by the compaction manager or unit tests.
```
This explanation doesn't explain to me how or why run_offstrategy_compaction() and perform_offstrategy_compaction() differ - they have the same signature.

bhalevy replied:
Maybe we should consider moving `run_offstrategy_compaction` to be private and befriend its callers.

That said, I talked with Raphael already about defining a formal abstraction of the table/compaction api so that compaction-manager and compaction would be provided with an object through which they would call back into the table (or test) code, rather than calling them directly on the table.  This way compaction will operate on an abstract container of sstables that could be a table object, but it could also be a TWCS bucket, for example, or just a bunch of sstables in a temporary directory for testing.

nyh reviewed (approved):
Looks mostly good to me, but I left some small comments/requests behind.

bhalevy commented:
In v5:
- fixed commitlog message re: compaction_manager::perform_offstrategy
- made table::run_offstrategy_compaction private
  - and renamed to do_run_offstrategy_compaction

avikivity commented:
What's the difference between offstrategy and major compaction?

If the difference between the current table shape and the strategy goal is large, then offstrategy will perform something similar to major compaction, no?

And if it's small, then the compaction strategy will work to reduce it.

I'm sure there is some difference, but does it merit a new API?

bhalevy commented:
> What's the difference between offstrategy and major compaction?
> 
> If the difference between the current table shape and the strategy goal is large, then offstrategy will perform something similar to major compaction, no?
> 
> And if it's small, then the compaction strategy will work to reduce it.
> 
> I'm sure there is some difference, but does it merit a new API?

Major compaction's goal is different than offstrategy compaction.
The former's goal, e.g. with STCS, is to compact all data into a single sstable,
while the latter's goal is to bring the sstables closer to the steady-state compaction-strategy invariant,
i.e. a state equivalent to after regular compaction (of all buckets).

nyh commented:
> Major compaction's goal is different than offstrategy compaction.
> The former's goal, e.g. with STCS, is to compact all data into a single sstable,
> while the latter's goal is to bring the sstables closer to the steady-state compaction-strategy invariant,
> i.e. a state equivalent to after regular compaction (of all buckets).

So one might say that "off-strategy" compaction is a confusing name - it's doesn't do something "off-strategy" but quite the opposite - it tries to return the current off-strategy situation to be more like the strategy wants. Maybe it should be called "back-to-strategy compaction" :-) In any case, we clearly need documentation for this feature - the one-line string in the REST API explains nothing - but I saw you opened an issue for that.

raphaelsc commented:
On Mon, Aug 9, 2021, 6:58 AM nyh ***@***.***> wrote:

> Major compaction's goal is different than offstrategy compaction.
> The former's goal, e.g. with STCS, is to compact all data into a single
> sstable,
> while the latter's goal is to bring the sstables closer to the
> steady-state compaction-strategy invariant,
> i.e. a state equivalent to after regular compaction (of all buckets).
>
> So one might say that "off-strategy" compaction is a confusing name - it's
> doesn't do something "off-strategy" but quite the opposite - it tries to
> return the current off-strategy situation to be more like the strategy
> wants. Maybe it should be called "back-to-strategy compaction" :-) In any
> case, we clearly need documentation for this feature - the one-line string
> in the REST API explains nothing - but I saw you opened an issue for that.
>
Actually it's called offstrategy because it's a compaction performed in a
maintainance sstable set where sstables in there were created by
maintenance operations (they aren't even considered by regular compaction
therefore offstrategy). The offstrategy compaction is not even driven by
compaction shares, maintenance shares is used instead. The offstrategy goal
is to prepare the maintenance sstables for integration into the main set.


—
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/9095#issuecomment-895095955>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKYA4Z3TJLHMOVZP6HYU4TT36REHANCNFSM5A75Q6XA>
> .
>

nyh commented:
> Actually it's called offstrategy because it's a compaction performed in a maintainance sstable set where sstables in there were created by maintenance operations (they aren't even considered by regular compaction therefore offstrategy).

Oh! This is quite different from what @bhalevy wrote above - are you saying that "off strategy compaction" doesn't even touch the "normal" sstables which major compaction works on - rather, it only works on some new sstables recently created by maintenance operations (which operations)? So if you didn't recently do any maintenance operations, it will not do anything at all?

raphaelsc commented:
On Mon, Aug 9, 2021, 7:33 AM nyh ***@***.***> wrote:

> Actually it's called offstrategy because it's a compaction performed in a
> maintainance sstable set where sstables in there were created by
> maintenance operations (they aren't even considered by regular compaction
> therefore offstrategy).
>
> Oh! This is quite different from what @bhalevy
> <https://github.com/bhalevy> wrote above - are you saying that "off
> strategy compaction" doesn't even touch the "normal" sstables which major
> compaction works on - rather, it only works on some new sstables recently
> created by maintenance operations (which operations)? So if you didn't
> recently do any maintenance operations, it will not do anything at all?
>

Correct

—
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/9095#issuecomment-895116862>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKYA42CGITMCR6JNHTSZADT36VHFANCNFSM5A75Q6XA>
> .
>

bhalevy commented:
> > Actually it's called offstrategy because it's a compaction performed in a maintainance sstable set where sstables in there were created by maintenance operations (they aren't even considered by regular compaction therefore offstrategy).
> 
> Oh! This is quite different from what @bhalevy wrote above - are you saying that "off strategy compaction" doesn't even touch the "normal" sstables which major compaction works on - rather, it only works on some new sstables recently created by maintenance operations (which operations)? So if you didn't recently do any maintenance operations, it will not do anything at all?

Yeah, this is better put this way.

bhalevy commented:
@nyh so what's the verdict?  Are you ok with merging this change?

bhalevy commented:
@avikivity please ack.  Are you ok with this change?

avikivity commented:
> > > Actually it's called offstrategy because it's a compaction performed in a maintainance sstable set where sstables in there were created by maintenance operations (they aren't even considered by regular compaction therefore offstrategy).
> > 
> > 
> > Oh! This is quite different from what @bhalevy wrote above - are you saying that "off strategy compaction" doesn't even touch the "normal" sstables which major compaction works on - rather, it only works on some new sstables recently created by maintenance operations (which operations)? So if you didn't recently do any maintenance operations, it will not do anything at all?
> 
> Yeah, this is better put this way.

I don't understand then how we can be in a situation that requires the API call. If we did a maintenance operation, it will automatically initiate offstrategy compaction into the main directory.

bhalevy commented:
> I don't understand then how we can be in a situation that requires the API call. If we did a maintenance operation, it will automatically initiate offstrategy compaction into the main directory.

It's not a requirement, just an optimization.
With regular repair, we don't know exactly when the meta-operation is finished since it is managed externally, e.g. by scylla-manager. Therefore, the decision was that we arm a 5 minutes timer that will submit the offstrategy compaction, if scylla is idle and no new repair operation starts.

This means that in the normal case we have a 5 minutes delay after repair finishes and until offstrategy compaction starts.
5 minutes is typically shorter than the time to run repair, but still it's a pretty long window, so the idea was to use this api to perform the offstrategy compaction immediately, when the external scheduler (e.g. scylla-manager, or a test) is done with repair and offstrategy compaction is good to go.

avikivity commented:
> > I don't understand then how we can be in a situation that requires the API call. If we did a maintenance operation, it will automatically initiate offstrategy compaction into the main directory.
> 
> It's not a requirement, just an optimization.
> With regular repair, we don't know exactly when the meta-operation is finished since it is managed externally, e.g. by scylla-manager. Therefore, the decision was that we arm a 5 minutes timer that will submit the offstrategy compaction, if scylla is idle and no new repair operation starts.
> 
> This means that in the normal case we have a 5 minutes delay after repair finishes and until offstrategy compaction starts.
> 5 minutes is typically shorter than the time to run repair, but still it's a pretty long window, so the idea was to use this api to perform the offstrategy compaction immediately, when the external scheduler (e.g. scylla-manager, or a test) is done with repair and offstrategy compaction is good to go.

So, this API call is to save five minutes?

bhalevy commented:
> > > I don't understand then how we can be in a situation that requires the API call. If we did a maintenance operation, it will automatically initiate offstrategy compaction into the main directory.
> > 
> > 
> > It's not a requirement, just an optimization.
> 
> This will need to be documented well - if it's at all useful... The first description I read in this patch made it sound like "off-strategy compaction" is a legitimate alternative to "major compaction", that just behaves a bit differently in a way I didn't understand. Now it has become clear that "off-strategy compaction" only useful after some unspecified "management operations" (which specific operations are those? repair? anything else?), but even that happens even without this command, so what this command apparently does is to **expedite** this compaction which will happen in 5 minutes anyway. None of this can be understood by the description of that command.
> 
> > With regular repair, we don't know exactly when the meta-operation is finished since it is managed externally, e.g. by scylla-manager. Therefore, the decision was that we arm a 5 minutes timer that will submit the offstrategy compaction, if scylla is idle and no new repair operation starts.
> 
> Maybe the default shouldn't be 5 minutes - that users might find a long boring wait, and look for a command to expedite this. What if, instead, it was a 1 minute delay? 1 minute is more than enough for scylla-manager to start another repair if it wants, but it's not long enough for users to go looking for a command to cancel that delay.
> 
> P.S. it's not that I think that there's anything "wrong" with this new API option, it's just that wouldn't it be better if it didn't have to exist, and users wouldn't even need to learn what it means?

@nyh First, I think that offstrategy compaction can be triggered during repair when there are enough sstables already accumulated.  By default we limit each compaction job to 32 sstables anyhow so there's no point in accumulating more than that per shard. Then, regarding 1 minute vs. 5 minutes, @asias and @raphaelsc what were these 5 minutes based on?

raphaelsc commented:
> > > > I don't understand then how we can be in a situation that requires the API call. If we did a maintenance operation, it will automatically initiate offstrategy compaction into the main directory.
> > > 
> > > 
> > > It's not a requirement, just an optimization.
> > 
> > 
> > This will need to be documented well - if it's at all useful... The first description I read in this patch made it sound like "off-strategy compaction" is a legitimate alternative to "major compaction", that just behaves a bit differently in a way I didn't understand. Now it has become clear that "off-strategy compaction" only useful after some unspecified "management operations" (which specific operations are those? repair? anything else?), but even that happens even without this command, so what this command apparently does is to **expedite** this compaction which will happen in 5 minutes anyway. None of this can be understood by the description of that command.
> > > With regular repair, we don't know exactly when the meta-operation is finished since it is managed externally, e.g. by scylla-manager. Therefore, the decision was that we arm a 5 minutes timer that will submit the offstrategy compaction, if scylla is idle and no new repair operation starts.
> > 
> > 
> > Maybe the default shouldn't be 5 minutes - that users might find a long boring wait, and look for a command to expedite this. What if, instead, it was a 1 minute delay? 1 minute is more than enough for scylla-manager to start another repair if it wants, but it's not long enough for users to go looking for a command to cancel that delay.
> > P.S. it's not that I think that there's anything "wrong" with this new API option, it's just that wouldn't it be better if it didn't have to exist, and users wouldn't even need to learn what it means?
> 
> @nyh First, I think that offstrategy compaction can be triggered during repair when there are enough sstables already accumulated. By default we limit each compaction job to 32 sstables anyhow so there's no point in accumulating more than that per shard. Then, regarding 1 minute vs. 5 minutes, @asias and @raphaelsc what were these 5 minutes based on?

@asias picked this. It's not a definitive value, but it's a reasonable one to begin with. Small values may lead to offstrategy running too early and that leads to inefficiency.  Even if offstrategy takes a couple of minutes to kick off, that's ok because data from non overlapping sstables are being efficiently served from maintenance set. It's better to delay offstrategy than running it too early.
